cranelift/codegen/src/egraph/cost.rs

```diff
@@ -43,12 +86,49 @@ impl Cost {
         Cost(0)
     }
 
+    fn new(opcode_cost: u32, depth: u8) -> Cost {
+        debug_assert!(
+            opcode_cost <= Self::MAX_OP_COST,
+            "Cost::new: given opcode cost of {opcode_cost} is larger than max of {}",
+            Self::MAX_OP_COST,
+        );
+        Cost((opcode_cost << Self::DEPTH_BITS) | u32::from(depth))
+    }
+
+    fn depth(&self) -> u8 {
+        let depth = self.0 & Self::DEPTH_MASK;
+        u8::try_from(depth).unwrap()
+    }
+
+    fn op_cost(&self) -> u32 {
+        (self.0 & Self::OP_COST_MASK) >> Self::DEPTH_BITS
```

Comment:
We could skip the mask here, as we're shifting off all the bits that get zeroed by the mask.
```suggestion
        self.0 >> Self::DEPTH_BITS
```

Reply:
Oh we could probably have an ISLE rewrite for this too!

```
(ishl (band a k1) k2) ==> (ishl k2) when shift_mask(k1) == k2 - 1
```

cc @afonso360: maybe I can nerd snipe you :)